grin/src/adapters.rs

```diff
@@ -398,16 +410,21 @@ impl NetToChainAdapter {
 	// TODO - currently only request block from a single peer
 	// consider additional peers for redundancy?
 	fn request_block(&self, bh: &BlockHeader, addr: &SocketAddr) {
-		if let None = wo(&self.peers).adapter.get_block(bh.hash()) {
+        self.request_block_by_hash(bh.hash(), addr)
+	}
+
+	fn request_block_by_hash(&self, h: Hash, addr: &SocketAddr) {
+		if let None = wo(&self.peers).adapter.get_block(h.clone()) {
```

Comment:
This loads a whole block, including parsing it, just to check if we already know about it. How about using `block_exists` instead of `get_block`?